File: examples/school/go/app/validation.go
package app

import (
	"encoding/json"
	"strconv"

	"github.com/cangrejometralleta/OneTwoThree/examples/school/go/faults"
	"github.com/cangrejometralleta/OneTwoThree/examples/school/go/school"
	"github.com/cangrejometralleta/OneTwoThree/examples/school/go/transport"
)

// A Request can be Malformed without Breaking a single Business Rule.
// Those Failures Belong to the Application.
var (
	ErrBodyIsBroken = faults.RefuseInvalidInput("body is not valid JSON")
	ErrPathIsBroken = faults.RefuseInvalidInput("path Holds no valid Identity")
)

// ReadPathNumber Pulls an Identity out of the Path.
func ReadPathNumber(req transport.Request) (uint64, error) {
	raw, err := strconv.ParseUint(req.Path["id"], 10, 64)
	if err != nil {
		return 0, ErrPathIsBroken
	}

	return raw, nil
}

// ReadJSONBody Decodes the Wire into whatever Shape the Caller Expects.
// It Validates the Form, never the Meaning: the Core Owns the Rules.
func ReadJSONBody[T any](req transport.Request) (T, error) {
	var body T

	if err := json.Unmarshal(req.Body, &body); err != nil {
		return body, ErrBodyIsBroken
	}

	return body, nil
}

// ReadPageRequest Reads Pagination, Defaulting to the whole Set.
func ReadPageRequest(req transport.Request) (school.Page, error) {
	page := school.Page{
		Number: ReadWholeOrRefuse(req.Query["page"]),
		Size:   ReadWholeOrRefuse(req.Query["size"]),
	}

	return page, page.CheckPageBounds()
}

// ReadWholeOrRefuse Reads a whole Number, or Returns one the Bounds Refuse.
// An absent Value Means the whole Set; a Word Means the Caller Mistyped.
// Swallowing the Parse Error would Answer two hundred to Nonsense.
func ReadWholeOrRefuse(value string) int {
	if value == "" {
		return 0
	}

	number, err := strconv.Atoi(value)
	if err != nil {
		return -1
	}

	return number
}

File: examples/school/java/src/main/java/com/example/app/Validation.java
package com.example.app;

import com.example.faults.Fault;
import com.example.school.Page;
import com.example.transport.Request;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;

/**
 * A Request can be Malformed without Breaking a single Business Rule.
 * Those Failures Belong to the Application.
 */
public final class Validation {

  private Validation() {
  }

  public static final Fault BODY_IS_BROKEN = Fault.refuseInvalidInput("body is not valid JSON");
  public static final Fault PATH_IS_BROKEN = Fault.refuseInvalidInput("path Holds no valid Identity");

  private static final ObjectMapper MAPPER = new ObjectMapper();

  /** ReadPathNumber Pulls an Identity out of the Path. */
  public static long readPathNumber(Request request) {
    try {
      long id = Long.parseLong(request.pathValue("id"));
      if (id < 0) {
        throw PATH_IS_BROKEN;
      }
      return id;
    } catch (NumberFormatException broken) {
      throw PATH_IS_BROKEN;
    }
  }

  /**
   * ReadJSONBody Decodes the Wire into whatever Shape the Caller Expects.
   * It Validates the Form, never the Meaning: the Core Owns the Rules.
   */
  public static <T> T readJsonBody(Request request, Class<T> shape) {
    try {
      T body = MAPPER.readValue(new String(request.body(), StandardCharsets.UTF_8), shape);
      if (body == null) {
        throw BODY_IS_BROKEN;
      }
      return body;
    } catch (Exception broken) {
      throw BODY_IS_BROKEN;
    }
  }

  /** ReadPageRequest Reads Pagination, Defaulting to the whole Set. */
  public static Page readPageRequest(Request request) {
    Page page = new Page(wholeOr(request.queryValue("page")), wholeOr(request.queryValue("size")));
    page.checkPageBounds();

    return page;
  }

  private static int wholeOr(String value) {
    try {
      return value.isBlank() ? 0 : Integer.parseInt(value);
    } catch (NumberFormatException notANumber) {
      return -1;
    }
  }
}
